### PR TITLE
[CDAP-20852] Publish AuditLogRequest to Ext Auth Service. This will contain more meta data for audit log publishing. 

### PR DESCRIPTION
**Pull request by sahusanket**

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java [hunk 1]
```diff
@@ -88,7 +91,11 @@ public void enforce(FullHttpRequest request, HttpResponder responder) throws Exc
           authorizationPrivilege.getPrincipal(),
           permissions);
     }
-    responder.sendStatus(HttpResponseStatus.OK);
+    Queue<AuditLogContext>  auditLogContextQueue = SecurityRequestContext.getAuditLogQueue();
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java:98]: You probably want same thing for `isVisible()`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/AuditLogSingleTopicSubscriberService.java [hunk 2]
```diff
@@ -130,29 +126,26 @@ protected void storeMessageId(StructuredTableContext context, String messageId)
...
+    List<AuditLogRequest> auditLogRequestList = new ArrayList<>();
 
     while (messages.hasNext()) {
-      ImmutablePair<String, AuditLogContext> next = messages.next();
-      AuditLogContext auditLogContext = next.getSecond();
-      if (auditLogContext.isAuditLoggingRequired()){
-        auditLogContextQueue.add(auditLogContext);
-      }
+      ImmutablePair<String, AuditLogRequest> next = messages.next();
+      auditLogRequestList.add(next.getSecond());
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/AuditLogSingleTopicSubscriberService.java]: nit: This looks like `Streams.stream(messages).map(ImmutablePair::getSecond).collect(toList())`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/AuditLogSingleTopicSubscriberService.java [hunk 3]
```diff
@@ -130,29 +126,26 @@ protected void storeMessageId(StructuredTableContext context, String messageId)
...
-    if (!auditLogContextQueue.isEmpty()) {
-      LOG.debug("Publishing a queue of Audit Log events of size {} events.", auditLogContextQueue.size());
+    if (!auditLogRequestList.isEmpty()) {
+      LOG.debug("Publishing a queue of Audit Log events of size {} events.", auditLogRequestList.size());
       AuditLoggerSpi.PublishStatus publishStatus =
-        this.accessControllerInstantiator.get().publishAuditLogs(auditLogContextQueue);
+        this.accessControllerInstantiator.get().publishAuditLogs(auditLogRequestList);
       // TODO : This logic can change based on how Auth Ext publishes a batch.
-      if (publishStatus.equals(AuditLoggerSpi.PublishStatus.UNSUCCESSFUL)) {
+      if (publishStatus == null || publishStatus.equals(AuditLoggerSpi.PublishStatus.UNSUCCESSFUL)) {
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/AuditLogSingleTopicSubscriberService.java]: It was already here, but you don't need equals for enums. I would just do `publishStatus != PUBLISHED`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/MessagingAuditLogWriter.java [hunk 4]
```diff
@@ -81,37 +82,33 @@ public MessagingAuditLogWriter(MessagingService messagingService,
...
-   * @param auditLogContexts
+   * pushes a log entry to respective messaging topic
    */
   @Override
-  public void publish(@Nullable Queue<AuditLogContext> auditLogContexts) throws IOException {
+  public void publish(@Nullable AuditLogRequest auditLogRequest) throws IOException {
 
-    if (auditLogContexts != null && auditLogContexts.isEmpty()){
+    if (auditLogRequest != null && auditLogRequest.getAuditLogContextQueue().isEmpty()
+      && auditLogRequest.getAuditLogContextQueue().stream().noneMatch(AuditLogContext::isAuditLoggingRequired)){
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/MessagingAuditLogWriter.java]: Do you mean `||`? Currently it checks for noneMatch in already checked empty collection

Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/MessagingAuditLogWriter.java]: I would also probably filter out the ones that `!isAuditLoggingRequired` much earlier. Otherwise if you have a mix, you would serialize all of those

File: cdap-common/src/main/java/io/cdap/cdap/common/metrics/MetricsReporterHook.java [hunk 5]
```diff
@@ -86,6 +89,9 @@ public boolean preCall(HttpRequest request, HttpResponder responder, HandlerInfo
 
   @Override
   public void postCall(HttpRequest request, HttpResponseStatus status, HandlerInfo handlerInfo) {
+
+    setAuditLogInfoInContext(request, status, handlerInfo);
```
Comment (tivv) [inline cdap-common/src/main/java/io/cdap/cdap/common/metrics/MetricsReporterHook.java]: I would either have it as a separate hook or move this hook out of metrics as this has nothing to do with metrics

File: cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/AuditLogSingleTopicSubscriberService.java (hunk 2 above)
Comment (sahusanket) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/AuditLogSingleTopicSubscriberService.java]: The "Streams" is coming from `org.apache.curator.shaded.com.google.common.collect` 
so we want to use it ?

Comment (sahusanket) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/AuditLogSingleTopicSubscriberService.java]: I tried, it gave a classloading error.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java [hunk 6]
```diff
@@ -71,6 +77,7 @@ public class RemotePrivilegesHandler extends AbstractRemoteSystemOpsHandler {
   @POST
   @Path("/enforce")
   public void enforce(FullHttpRequest request, HttpResponder responder) throws Exception {
+    LOG.warn("SANKET_TEST : isVisible : thread - {}", Thread.currentThread().getName());
```
Comment (itsankit-google) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java]: nit: please remove testing logs.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java [hunk 7]
```diff
@@ -88,21 +95,39 @@ public void enforce(FullHttpRequest request, HttpResponder responder) throws Exc
...
+    //This should be written by the Client who is calling this service.
+    SecurityRequestContext.clearAuditLogQueue();
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(auditLogContextQueue));
   }
 
   @POST
   @Path("/isVisible")
+  @SuppressWarnings("unchecked")
   public void isVisible(FullHttpRequest request, HttpResponder responder) throws Exception {
+    LOG.warn("SANKET_TEST : isVisible : thread - {}", Thread.currentThread().getName());
```
Comment (itsankit-google) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java]: similar comment here

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java [hunk 8]
```diff
@@ -88,21 +95,39 @@ public void enforce(FullHttpRequest request, HttpResponder responder) throws Exc
...
-    Set<? extends EntityId> visiableEntities = accessEnforcer.isVisible(entityIds, principal);
-    LOG.trace("Returning entities visible for principal {} as {}", principal, visiableEntities);
-    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(visiableEntities));
+    Set<? extends EntityId> visibleEntities = accessEnforcer.isVisible(entityIds, principal);
+    Map<? extends EntityId, AuthorizationResponse> entityToAuthResponseMap =
+      SecurityRequestContext.getEntityToAuthResponseMap();
+    SecurityRequestContext.clearEntityToAuthResponseMap();
+    //For internal checks, no audit log required, so it is possible that map from SecurityRequestContext will be empty
+    // Or might contain less no of visible entities.
+    Map<? extends EntityId, AuthorizationResponse> entityToAuthResponseMapFull =visibleEntities.stream()
```
Comment (itsankit-google) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java]: checkstyle: `entityToAuthResponseMapFull = visibleEntities.stream()`

File: cdap-security/src/main/java/io/cdap/cdap/security/authorization/RemoteAccessEnforcer.java [hunk 9]
```diff
@@ -220,6 +255,7 @@ public void enforceOnParent(EntityType entityType, EntityId parentId, Principal
   @Override
   public Set<? extends EntityId> isVisible(Set<? extends EntityId> entityIds, Principal principal)
       throws AccessException {
+    LOG.warn("SANKET_TEST : isVisible : thread - {}", Thread.currentThread().getName());
```
Comment (itsankit-google) [inline cdap-security/src/main/java/io/cdap/cdap/security/authorization/RemoteAccessEnforcer.java]: same comment here

File: cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authentication/SecurityRequestContext.java [hunk 10]
```diff
@@ -109,12 +139,19 @@ public static void reset() {
...
     auditLogContextQueue.remove();
+    auditLogRequest.remove();
+    entityToAuthResponseMap.remove();
   }
 
   /**
    * Creates a queue if not present and adds the {@link AuditLogContext} to it.
    */
   public static void enqueueAuditLogContext(AuditLogContext auditLog) {
+
```
Comment (itsankit-google) [inline cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authentication/SecurityRequestContext.java]: nit: remove empty line

**Review by itsankit-google — APPROVED** [review]

File: cdap-security/src/main/java/io/cdap/cdap/security/authorization/RemoteAccessEnforcer.java [hunk 11]
```diff
@@ -158,21 +179,22 @@ public EnforcementResponse load(AuthorizationPrivilege authorizationPrivilege)
...
 
           @Override
-          public Map<VisibilityKey, Boolean> loadAll(Iterable<? extends VisibilityKey> keys)
+          public Map<VisibilityKey, VisibilityValue> loadAll(Iterable<? extends VisibilityKey> keys)
               throws Exception {
             LOG.trace("Cache miss for {}", keys);
             return loadVisibility(keys);
           }
         });
+
```
Comment (itsankit-google) [inline cdap-security/src/main/java/io/cdap/cdap/security/authorization/RemoteAccessEnforcer.java]: nit: remove empty line

File: cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/AuditLogSingleTopicSubscriberService.java [hunk 12]
```diff
@@ -130,29 +126,22 @@ protected void storeMessageId(StructuredTableContext context, String messageId)
    */
   @Override
   protected void processMessages(StructuredTableContext structuredTableContext,
-                                 Iterator<ImmutablePair<String, AuditLogContext>> messages) throws Exception {
+                                 Iterator<ImmutablePair<String, AuditLogRequest>> messages) throws Exception {
 
-    Queue<AuditLogContext> auditLogContextQueue = new ArrayDeque<>();
+    List<AuditLogRequest> auditLogRequestList = new ArrayList<>();
```
Comment (itsankit-google) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/AuditLogSingleTopicSubscriberService.java]: where is this list used?

File: cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuditLogRequest.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+ */
+public class AuditLogRequest {
+
+  private final int operationResponseCode;
+  private final String uri;
+  private final String handler;
+  private final String method;
+  private final String methodType;
+  private final Queue<AuditLogContext> auditLogContextQueue;
+  private final Long startTime;
```
Comment (itsankit-google) [inline cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuditLogRequest.java]: please add `time unit` in variable name.

For example: `startTimeMillis`

File: cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuditLogRequest.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+  private final Long endTime;
```
Comment (itsankit-google) [inline cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuditLogRequest.java]: similar comment here

File: cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuditLoggerSpi.java [hunk 15]
```diff
@@ -31,7 +33,17 @@ public interface AuditLoggerSpi {
...
-    UNSUCCESSFUL
+    UNSUCCESSFUL;
+
+    private Exception ex;
+
+    public Exception getEx() {
+      return ex;
+    }
+
+    public void setEx(Exception ex) {
```
Comment (itsankit-google) [inline cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuditLoggerSpi.java:44]: it should return `PublishStatus`?

File: cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuditLogRequest.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.util.Queue;
+import javax.annotation.Nullable;
+
+/**
+ * The object that will be sent to Auth Extension which would contain the collections of audit logs
+ * {@link AuditLogContext} related to a Single Call or Operation.
+ * This also contains various other metadata required for Audit Logging.
+ */
+public class AuditLogRequest {
```
Comment (itsankit-google) [inline cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuditLogRequest.java:27]: Please also add `userIp` in `AuditLogRequest`. Should be feasible to get from `SecurityRequestContext`.

File: cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuditLoggerSpi.java (hunk 15 above)
Comment (sahusanket) [reply cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuditLoggerSpi.java:44]: why?

Comment (itsankit-google) [reply cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuditLoggerSpi.java:44]: nvm, resolving the comment.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java [hunk 17]
```diff
@@ -160,7 +160,8 @@ protected void doShutdown() throws Exception {
     RetryStrategy retryStrategy = RetryStrategies.timeLimit(
         gracefulShutdownMillis,
         TimeUnit.MILLISECONDS,
-        getRetryStrategy());
+        getRetryStrategy()); //we are passing in constru...SUPER  ...exponential
```
Comment (itsankit-google) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java:163]: why this change?
